src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderValidationTests.cs

```diff
@@ -239,6 +239,84 @@ public void GetService_DoesNotThrow_WhenGetServiceForNonScopedImplementationWith
             Assert.NotNull(result);
         }
 
+        [Fact]
+        public void BuildServiceProvider_ValidateOnBuild_Throws_WhenScopedIsInjectedIntoSingleton()
+        {
+            // Arrange
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddScoped<IBar, Bar>();
+            serviceCollection.AddSingleton<IFoo, Foo>();
+
+            // Act + Assert
+            var aggregateException = Assert.Throws<AggregateException>(() => serviceCollection.BuildServiceProvider(new ServiceProviderOptions() { ValidateOnBuild = true, ValidateScopes = true }));
```

Comment:
I assume without the fix here, that no exception was thrown here.

Reply:
An exception was thrown here, but not in the case with the reversed order. The aspnet core repo happens to test the reverse order scenario. This happened because in the previous code, the scoped-in-singleton check was not performed if a callsite was already cached. See here as well: https://github.com/dotnet/runtime/issues/98551#issuecomment-1952177976